PWGPP-570: Merging ionTailMerge branch into master

miranov25 opened the pull request:
Please do not approve - I prefer to run a test first 

# Modifications:

* ion tail _ common mode changes related modification in November-March

### After merging with master:
* Debug streaming - down-sampling of cluster  - used for tuning of the cluster error parameterization
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239113&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239113
  * Ion tail - integral scaling 
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239114&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239114 
* Write local baseline shift due ion tail and common mode to the clusters:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239115&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239115
* Multiplicity and dEdx dependent Cluster error parameterization - IMPLEMETED (22.11.2015) see
  
  * https://indico.cern.ch/event/790434/contributions/3647824/attachments/1949646/3235806/PWGPP-570_ErrorParameterizationAndClusterAssciation.pdf 
https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239116&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239116
* Switch to enable dEdx, multiplcity dependent parameterization:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239129&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239129
* Usage of the missing cluster correction in the dEdx calculation -  IMPLEMETED (22.11.2015) see:
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239117&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239117
  * https://alice.its.cern.ch/jira/browse/PWGPP-570?focusedCommentId=239150&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239150

alibuild commented:
9a3e83fa069a2ae2a4c770ce788805853063c0fe: approval required: 1 of @davidrohr (David Rohr), @ekryshen (Evgeny Kryshen), @gconesab (Gustavo Conesa Balbastre), @wiechula (Jens Wiechula), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @mvala (Martin Vala), @chiarazampolli (Chiara Zampolli), @pantonioli (Pietro Antonioli), @pchrista (Panos Christakoglou), @peressounko (Dmitri Peresunko), @zconesa (Zaida Conesa Del Valle), @mfasDa (Markus Fasel), @aortizve (Antonio Ortiz Velasquez), @deepathoms (Deepa Thomas), @lcunquei (Leticia Cunqueiro Mendez), @qgp (Jochen Klein), @akalweit (Alexander Philipp Kalweit), @ddobrigk (David Dobrigkeit Chinellato), @shahor02 (Ruben Shahoyan); 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @pzhristov (Peter Hristov), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @chiarazampolli (Chiara Zampolli), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan); 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @qgp (Jochen Klein), @dberzano (Dario Berzano), @miranov25 (Marian Ivanov), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/1070) and comment at the bottom of the page.**_

miranov25 commented:
+test

alibuild commented:
9a3e83fa069a2ae2a4c770ce788805853063c0fe: approval required: 1 of @davidrohr (David Rohr), @ekryshen (Evgeny Kryshen), @gconesab (Gustavo Conesa Balbastre), @wiechula (Jens Wiechula), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @mvala (Martin Vala), @chiarazampolli (Chiara Zampolli), @pantonioli (Pietro Antonioli), @pchrista (Panos Christakoglou), @peressounko (Dmitri Peresunko), @zconesa (Zaida Conesa Del Valle), @mfasDa (Markus Fasel), @aortizve (Antonio Ortiz Velasquez), @deepathoms (Deepa Thomas), @lcunquei (Leticia Cunqueiro Mendez), @qgp (Jochen Klein), @akalweit (Alexander Philipp Kalweit), @ddobrigk (David Dobrigkeit Chinellato), @shahor02 (Ruben Shahoyan); 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @pzhristov (Peter Hristov), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @chiarazampolli (Chiara Zampolli), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/1070) and comment at the bottom of the page.**_

shahor02 commented:
+test

alibuild commented:
9a3e83fa069a2ae2a4c770ce788805853063c0fe: testing approved: will not be automatically merged; starting testing. If testing succeeds, merging will require further approval from 1 of @davidrohr (David Rohr), @ekryshen (Evgeny Kryshen), @gconesab (Gustavo Conesa Balbastre), @wiechula (Jens Wiechula), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @mvala (Martin Vala), @chiarazampolli (Chiara Zampolli), @pantonioli (Pietro Antonioli), @pchrista (Panos Christakoglou), @peressounko (Dmitri Peresunko), @zconesa (Zaida Conesa Del Valle), @mfasDa (Markus Fasel), @aortizve (Antonio Ortiz Velasquez), @deepathoms (Deepa Thomas), @lcunquei (Leticia Cunqueiro Mendez), @qgp (Jochen Klein), @akalweit (Alexander Philipp Kalweit), @ddobrigk (David Dobrigkeit Chinellato), @shahor02 (Ruben Shahoyan); 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @pzhristov (Peter Hristov), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @chiarazampolli (Chiara Zampolli), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

alibuild commented:
Error while checking build/AliRoot/release for 9a3e83fa069a2ae2a4c770ce788805853063c0fe:
```

```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/alisw/AliRoot/1070/9a3e83fa069a2ae2a4c770ce788805853063c0fe/build_AliRoot_release/fullLog.txt).

shahor02 commented:
@miranov25 I suspect your master was not rebased to remote: I see in the list of your already merged commits by others, like https://github.com/alisw/AliRoot/pull/1070/commits /9a3e83fa069a2ae2a4c770ce788805853063c0fe from Peter

fprino commented:
Hi Marian, all,

I see modifications in AliAnalysisManager.cxx and AliGenPythiaPlus that are not related to the TPC development.
I think this is related to the 2 latest commits in the list of this PR, namely
"fPythia object not persistent to avoid default constructur be called …" 
"Return kTRUE since kFALSE indicates to Root v6-18-04 real problems"
which are modifications done by Peter and Andreas.
Can you please check?

Ciao, Francesco

alibuild commented:
bd0428e48a125b043affa776e9a3ad50802d577a: approval required: 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @qgp (Jochen Klein), @dberzano (Dario Berzano), @miranov25 (Marian Ivanov), @shahor02 (Ruben Shahoyan); 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @pzhristov (Peter Hristov), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @chiarazampolli (Chiara Zampolli), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/1070) and comment at the bottom of the page.**_

miranov25 commented:
Hello Reuben and Francesco. 

Not clear what went wrong while merging in previous attempt.
I m almost sure I update my master before merging
To be sure I started again from clean master.
Could you test again?

Marian
```
[AliPhysics/latest] ~/software/aliceGSI/AliRoot $>  git merge ionTailMerge
Merge made by the 'recursive' strategy.
 STAT/AliTMinuitToolkit.cxx         |  20 +++--
 STAT/AliTMinuitToolkit.h           |   2 +-
 STEER/ESD/AliESDUtils.cxx          |   6 +-
 TPC/TPCbase/AliTPCRecoParam.cxx    |  12 ++-
 TPC/TPCbase/AliTPCRecoParam.h      |  25 +++++-
 TPC/TPCbase/AliTPCcalibDB.cxx      | 132 ++++++++++++++++++------------
 TPC/TPCbase/AliTPCclusterMI.cxx    |  10 ++-
 TPC/TPCbase/AliTPCclusterMI.h      |  11 ++-
 TPC/TPCrec/AliTPCReconstructor.cxx |   1 +
 TPC/TPCrec/AliTPCReconstructor.h   |   5 +-
 TPC/TPCrec/AliTPCseed.cxx          |  13 ++-
 TPC/TPCrec/AliTPCtracker.cxx       | 650 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
 TPC/TPCrec/AliTPCtracker.h         |   1 +
 TPC/TPCsim/AliTPCDigitizer.cxx     | 300 ++++++++++++++++++++++++++++++++++++-------------------------------
 test/configuration/configGSI.sh    |   2 +-
 15 files changed, 760 insertions(+), 430 deletions(-)
```

chiarazampolli commented:
+test

alibuild commented:
bd0428e48a125b043affa776e9a3ad50802d577a: approval required: 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @qgp (Jochen Klein), @dberzano (Dario Berzano), @miranov25 (Marian Ivanov), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/1070) and comment at the bottom of the page.**_

miranov25 commented:
+test

alibuild commented:
bd0428e48a125b043affa776e9a3ad50802d577a: testing approved: will not be automatically merged; starting testing. If testing succeeds, merging will require further approval from 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @qgp (Jochen Klein), @dberzano (Dario Berzano), @miranov25 (Marian Ivanov), @shahor02 (Ruben Shahoyan)

alibuild commented:
Error while checking build/AliRoot/release for bd0428e48a125b043affa776e9a3ad50802d577a:
```

```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/alisw/AliRoot/1070/bd0428e48a125b043affa776e9a3ad50802d577a/build_AliRoot_release/fullLog.txt).

shahor02 commented:
now the extra commits are gone. I suspect the failure is spurious, let's wait for the macos.

shahor02 commented:
@ktf, @MichaelLettrich CI reports a failure for release (immidiately after the push) w/o providing the log, is this this a CI problem?

ktf commented:
@MichaelLettrich did you restart this worker the other day? I just did it again.

ktf commented:
(regardless of that, looks like there are conflict when merging...)

alibuild commented:
bd0428e48a125b043affa776e9a3ad50802d577a: tests OK, approval required for merging: 1 of @davidrohr (David Rohr), @jgrosseo (Jan Fiete Grosse-Oetringhaus), @qgp (Jochen Klein), @dberzano (Dario Berzano), @miranov25 (Marian Ivanov), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to merge. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/1070) and comment at the bottom of the page.**_

ktf commented:
Checks are now successful, but we cannot merge...

shahor02 commented on TPC/TPCsim/AliTPCDigitizer.cxx:
```diff
@@ -704,149 +698,150 @@ void AliTPCDigitizer::DigitizeWithTailAndCrossTalk(Option_t* option)
+    }
+    timeResFunc.AddAt(timeResArr, isec); // Fill all trfs into a single TObjArray
+    nIonTailBins = graphRes[0]->GetN();
+    delete trfIndexArr;
```
@miranov25 I see here you have overwritten the memory leak I've fixed some time ago.

miranov25 replied:
I merged this part in March. I assumed I did it correctly. 
I will rebase again ...

shahor02 commented:
@ktf thanks, how did you know that there are conflicts (with the github saying there are none)?
In fact I am able to do
``git merge  miranov25/master`` on the up to date master w/o conflicts, but his original ionTailMerge produces lot of conflicts when being rebased on master (including the conflict with my fix for the memory leak).
@miranov25 I gues the best if you close this PR, do
````
git checkout master
git fetch origin && git reset --hard origin/master 
git checkout ionTailMerge
git rebase master
# solve conflicts
git push -f miranov25 master
```` 
then open PR for ionTailMerge

Cheers,
 Ruben

ktf commented:
I misread the message from Github GUI... It says "Unable to **rebase**". 

I can merge directly without rebasing, if that's ok, but I think in general it is good practice to rebase before merging..

shahor02 commented:
I also think it is better to rebase, right now there is at least 1 memory leak fix reverted.

miranov25 commented:
OK.
I will  
```
git checkout master
git fetch origin && git reset --hard origin/master 
git checkout ionTailMerge
git rebase master
# solve conflicts
git push -f miranov25 master
```
and open new pull request

miranov25 commented:
f...
meld is crashing in merging.
f...
```
Normal merge conflict for 'TPC/TPCsim/AliTPCDigitizer.cxx':
  {local}: modified file
  {remote}: modified file
Hit return to start merge resolution tool (meld): bc
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/meld/meldwindow.py", line 309, in on_delete_event
    response = page.on_delete_event()
  File "/usr/lib/python3/dist-packages/meld/filediff.py", line 842, in on_delete_event
    del self._cached_match
AttributeError: _cached_match
```

miranov25 commented:
using clion for resolving conflics  I could merge  - but automatic git merging ignored important modifications from iontailMerge
default merge removed my changes in:
````
      modified:   TPC/TPCbase/AliTPCcalibDB.cxx
        modified:   TPC/TPCrec/AliTPCtracker.cxx
        modified:   TPC/TPCsim/AliTPCDigitizer.cxx
````
one more time ...

miranov25 commented:
# Summarizing:

* git rebase master
  * my modifications in the AliTPCcalibDB and AliTPCtracker are ignored
* git merge master
  * Ruben modifications from AliTPCdigitizer are ignored

miranov25 commented:
We are stuck here - and I will have to leave soon.
My proposed solution use git merge master, and merge Ruben diffs later.
Do you agree?

shahor02 commented:
@miranov25 The memory leak you have reintroduced was relevant for the MC only and was not critical, though there might be others. But the proper solution would be to save your current files where your modifications as lost (AliTPCcalibDB and AliTPCtracker only), rebase to master, then manually fix the diffs between saved and rebased version of files.

If this cannot be done now, for the moment we probably can use merge, but as far as I can see, it will but much more difficult to track down later the fixes lost in the merge: you will need to rebase your branch to current master (f58237542c), resolve the conflicts, edit the diffs between the files where there were conflicts and commit them as new patches.

miranov25 commented:
@shahor02 

I checked both approaches - storing and diffing commits. 
In current situation, git rebase can not be used. All my recent commits, which I described in the JIRA PWGPP-570 are lost (e.g for the TPCrec directory - everything  after 15.10)
````
[AliPhysics/latest] ~/software/aliceGSI/AliRoot/TPC/TPCrec $> wc  *gitlog
  1370   5533  48920 TPCrecMerge.gitlog
  1370   5533  48920 TPCrecOrig.gitlog
  1189   4720  42225 TPCrecrebase.gitlog
````
This time, we have to use merge - and add manual fixes later.

shahor02 commented:
> PWGPP-570 are lost (e.g for the TPCrec directory - everything after 15.10)
I don't understand how this is possible: the last commit in the master: TPC/TPCrec is from 2018, rebase should not override any of your recent commits. Let me try

miranov25 commented:
in case you work on that, I merged AliTPCDigitizer already several times
here it is clear how to solve problems. This part i can provide.

shahor02 commented:
@miranov25 Could you check the branch https://github.com/shahor02/AliRoot/tree/mi_IT, it is your ionTailMerge rebased to master and some conflicts solved. I see all you recent commits in TPCrec.
The branch merges to master w/o problem.
If it is OK, just close this PR and open new one with mi_IT

miranov25 commented:
Merge as you prepared is OK
Closing this request
